File: BE/BaiTap-RS/src/main/java/com/JavaTraining/BaiTap_RS/teacher/domain/entity/Teacher.java
package com.JavaTraining.BaiTap_RS.teacher.domain.entity;

import java.time.LocalDate;
import java.time.LocalDateTime;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@NoArgsConstructor(access = lombok.AccessLevel.PROTECTED)
@Entity
@Table(
        name = "teacher",
        uniqueConstraints = {
                @UniqueConstraint(name = "uk_teacher_user", columnNames = "user_id"),
                @UniqueConstraint(name = "uk_teacher_code", columnNames = "teacher_code")
        })
public class Teacher {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "teacher_id", nullable = false)
    private Long id;

    @Column(name = "user_id")
    private Long userId;

    @Column(name = "teacher_code", nullable = false, length = 50)
    private String teacherCode;

    @Column(name = "teacher_name", nullable = false, length = 150)
    private String teacherName;

    @Column(name = "date_of_birth")
    private LocalDate dateOfBirth;

    @Column(name = "gender", length = 20)
    private String gender;

    @Column(name = "phone", length = 30)
    private String phone;

    @Column(name = "email", length = 150)
    private String email;

    @Column(name = "department", length = 100)
    private String department;

    @Column(name = "join_date")
    private LocalDate joinDate;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 20)
    private TeacherStatus status;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    public Teacher(
            Long userId,
            String teacherCode,
            String teacherName,
            LocalDate dateOfBirth,
            String gender,
            String phone,
            String email,
            String department,
            LocalDate joinDate,
            TeacherStatus status) {
        this.userId = userId;
        this.teacherCode = teacherCode;
        this.teacherName = teacherName;
        this.dateOfBirth = dateOfBirth;
        this.gender = gender;
        this.phone = phone;
        this.email = email;
        this.department = department;
        this.joinDate = joinDate;
        this.status = status;
    }

    @PrePersist
    /* default */ void onCreate() {
        LocalDateTime now = LocalDateTime.now();
        createdAt = now;
        updatedAt = now;
    }

    @PreUpdate
    /* default */ void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: FE/src/types/teacher.ts
export type TeacherStatus = 'ACTIVE' | 'ON_LEAVE' | 'INACTIVE'

export interface Teacher {
  id: number
  userId: number | null
  teacherCode: string
  teacherName: string
  dateOfBirth: string | null
  gender: string | null
  phone: string | null
  email: string | null
  department: string | null
  joinDate: string | null
  status: TeacherStatus
}

export interface TeacherFormValues {
  userId: number | null
  teacherCode: string
  teacherName: string
  dateOfBirth: string
  gender: string
  phone: string
  email: string
  department: string
  joinDate: string
  status: TeacherStatus
}

export interface CreateTeacherRequest {
  userId: number | null
  teacherCode: string
  teacherName: string
  dateOfBirth: string | null
  gender: string | null
  phone: string | null
  email: string | null
  department: string | null
  joinDate: string | null
  status: TeacherStatus
}

export type UpdateTeacherRequest = CreateTeacherRequest
